tests/object_file_system_test.php

```diff
@@ -585,6 +585,39 @@ public function test_can_generate_signed_url_by_hash_if_object_is_external() {
         }
     }
 
+    public function test_can_generate_signed_url_with_headers() {
+        $this->filesystem = new test_file_system();
+        $file = $this->create_remote_file();
+        $filehash = $file->get_contenthash();
+        try {
+            $headers = [
+                'Content-Disposition' => 'attachment; filename="filename.txt"',
+                'Content-Type' => 'text/plain',
+            ];
+            $signedurl = $this->filesystem->externalclient->generate_presigned_url($filehash, $headers);
+            $this->assertTrue($this->is_externally_readable_by_url($signedurl));
+        } catch (\coding_exception $e) {
+            $this->assertEquals($e->a, 'Pre-signed URLs not supported');
+        }
+    }
+
+    public function test_can_generate_signed_url_with_unicode_filename() {
+        $this->filesystem = new test_file_system();
+        $file = $this->create_remote_file();
+        $filehash = $file->get_contenthash();
+        try {
+            $headers = [
+                    'Content-Disposition' => 'attachment; filename="😀.txt"',
```

Comment:
Hello Everyone,
just lurking here 👀.
Is there any reason not to adopt even the [RFC 5987](https://www.rfc-editor.org/rfc/rfc5987#section-3.2) and [RFC 6266](https://www.rfc-editor.org/rfc/rfc6266#section-4.1) indications i.e.:
```
Content-Disposition: attachment; filename*=utf-8''%e2%82%ac%20rates
```
where the _filename_ there is something like `rawurlencode($originalfilename)`?
I mean something like:
```php
$result['filename'] = 'filename="' . utf8_encode($originalfilename) . '"'
                  . "; filename*=utf-8''" . rawurlencode($originalfilename);
```
Maybe it's not required by your use cases right now but it could be future proof.

HTH,
Matteo

Reply:
Hi Matteo, you're right, that probably would make sense. With Moodle we can safely assume that the filename is UTF-8 encoded, so specifying that in the header would be correct. 

We are currently doing the rawurlencode() seperately, which doesn't quite follow the pattern you indicated. I will do an additional commit to rework it as you suggested.

Reply:
> We should be OK just setting the filename as UTF-8 and doing `rawurlencode()`.

Modern browsers should be fine with that, though e.g. https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition is not clear enough on the back-compat.
Old benchmarks like http://test.greenbytes.de/tech/tc2231/ give some hints but I think its fine to tell the browsers are now running fine in the modern days 😉.

HTH,
Matteo